requirements/specifications/openrpc-extensions/app-passthrough-apis.md

```diff
@@ -0,0 +1,533 @@
+
+```json
+{
+    "methods": [
+        {
+            "name": "Keyboard.standard",
+            "tags": [
+                {
+                    "name": "capabilities",
+                    "x-provided-by": "Keyboard.onRequestStandard",
+                    "x-uses": [
+                        "xrn:firebolt:capability:input:keyboard"
+                    ]
+                }
+            ]
+        }
+    ]
+}
+```
+
+The method denoted by `x-provided-by` is referred to as the "*provider*" or "*provider method*" for the remainder of this document.
+
+The method with the `x-provided-by` extension is referred to as the "*platform method*" for the remainder of this document.
+
+To prevent unresolvable chaining of methods the `x-provided-by` extension **MUST NOT** be used on a method with any value in the `x-provides` extension.
+
+To prevent compound methods a platform method **MUST** `use` a single capability or `manage` a single capability, but not both.
+
+The provider method **MUST** provide the same capability that the platform method either uses or manages.
+
+If an platform method has no provider method then it is not a valid Firebolt OpenRPC method schema, and a validation error **MUST** be generated.
+
+### 3.2. Multiple Providers Extension
+Firebolt OpenRPC **MUST** support a `string` `x-multiple-providers` extension property on the `capabilities` tag that denotes a single method request may be provided by multiple apps on the device registering for the specified provider API, e.g.:
+
+```json
+{
+    "methods": [
+        {
+            "name": "Content.search",
+            "tags": [
+                {
+                    "name": "capabilities",
+                    "x-provided-by": "Discover.onRequestSearch",
+                    "x-multiple-providers": true,
+                    "x-uses": [
+                        "xrn:firebolt:capability:discovery:search"
+                    ]
+                }
+            ]
+        }
+    ]
+}
+```
+
+Setting `x-multiple-providers` to `true` means that all available apps that can provide the capability **MUST** be called and their results aggregated into an array for the final result.
+
+A platform method with `x-multiple-providers` set to `true` **MUST** have an `array` result type.
+
+## 4. Routing App pass-through APIs
+App pass-through APIs may be routed in one of several ways.
+
+When an app calls a platform method, i.e. one with an `x-provided-by` extension, the platform **MUST** use one of the routing methods defined in this section based on various properties of the method.
+
+### 4.1. No available providers
+When an app calls a platform method with an `x-provided-by` extension, the platform **MUST** return an unavailable error if there is no [candidate app](#7-provider-candidates) to execute the provider method.
+
+```json
+{
+    "id": 1,
+    "error": {
+        "code": -50300,
+        "message": "Capability is unavailable."
```

Comment:
should it be "xrn:firebolt:capability:discovery:search is unavailable" ?

Reply:
Wanted to sync up on this.

Do we want dynamic error messages?

Reply:
I would lean yes

Reply:
okay, I'll just have to write regex's for the errors when we finally put them in the OpenRPC. That was the main reason I kept the simple